### DISCUSSION TITLE
Flatbuffers & SDKs standardization

### DISCUSSION THREAD

**Discussion by DadiBit** [open-meteo/open-meteo#580] [General]
Greetings,
1. the original issue was [open-meteo/open-meteo-website #40](https://github.com/open-meteo/open-meteo-website/issues/40). I'm closing it and shift it here, as a discussion on the main repo.
2. I'm currently developing the Kotlin Multiplatform SDK, which would support not only jvm, but js, linux, windows, macos, ios, android and wasm too: I would like to propose going "all-in" with KMP, as it would reduce the codebase and I feel like it would be much easier to integrate. A simple gh action that tests PRs with `./gradlew macosTest` should suffice.
3. While developing the SDK I experimented with git submodules, and would like to propose to create a dedicated repo to host flatbuffer schema. Advantage: external SDK repos (or server repo, for that matter) can import just the `.fbs` files with `git submodule` + manually update them.
4. @patrick-zippenfenig could we bring back [6facc70](https://github.com/open-meteo/sdk/commit/6facc704ed04872f69c56369c1fe83f8ca9e5e67) fbs files?
5. I was also thinking about splitting the code a bit, creating some separate files for common stuff like `Units`, wdyt?
6. If needed, I do have some fbs for the geocoding API :smile:

**Comment by patrick-zippenfenig** [comment]
> @patrick-zippenfenig could we bring back [6facc70](https://github.com/open-meteo/sdk/commit/6facc704ed04872f69c56369c1fe83f8ca9e5e67) fbs files?

They are still there, but only one fbs is required for now: https://github.com/open-meteo/sdk/tree/main/flatbuffers

> I was also thinking about splitting the code a bit, creating some separate files for common stuff like Units, wdyt?
If needed, I do have some fbs for the geocoding API 😄
> If needed, I do have some fbs for the geocoding API 

Yes, there will be more FlatBuffers Schema files for other APIs (elevation, geocoding, geocoding-database, weather-database, measurements?). Currently, I am not sure if it makes sense to split `Units` and other enums. Having just a couple of FBS files for contained APIs might be better.

It may take a couple of weeks until the geocoding API will support FlatBuffers. I am currently integrating more weather datasets and want to continue working on a open-data distribution.

**Comment by DadiBit** [comment]
> Yes, there will be more FlatBuffers Schema files for other APIs (elevation, geocoding, geocoding-database, weather-database, measurements?). Currently, I am not sure if it makes sense to split `Units` and other enums. Having just a couple of FBS files for contained APIs might be better.

In order to keep a more organized git history I think keeping those common enums separate would make sens. After all, `Units` appear in multiple APIs (forecast, marine, air quality...).
Perhaps something like:
```
flatbuffer-schema
 +- airquality.fbs
 +- elevation.fbs
 +- ensemble.fbs
 +- flood.fbs
 +- forecast.fbs
 +- geocoding.fbs
 +- historical.fbs
 +- marine.fbs
 +- aggregation.fbs
 +- unit.fbs
```

> It may take a couple of weeks until the geocoding API will support FlatBuffers. I am currently integrating more weather datasets and want to continue working on a open-data distribution.

Great! In the meantime, if it can help, here's my "translation" of the proto source (I couldn't use `flatc --proto`, since the entry with `id: 3` was missing):
```
table GetResponse {
  id: int;
  name: string;
  _3: string; // maybe remove this, idk
  latitude: float;
  longitude: float;
  ranking: float;
  elevation: float;
  feature_code: string;
  country_code: string;
  admin1_id: int;
  admin2_id: int;
  admin3_id: int;
  admin4_id: int;
  timezone: string;
  population: int;
  alternative_names: [string];
  postcodes: [string];
  country_id: int;
  country: int;
  admin1: string = "";
  admin2: string = "";
  admin3: string = "";
  admin4: string = "";
}

table SearchResponse {
  results: [GetResponse];
  generation_time_milliseconds: float;
}
```

Still lacking some enums, tho.

**Reply by DadiBit** [reply]
@patrick-zippenfenig If you want to reduce network footprint by ~20 bytes (better than nothing) by increasing computational usage, here's an idea:

- Let's say we have a timeframe of 20 seconds from 100 to 300 Unix-epoch seconds; in other words: 100, 120, 140, 160, 180 ... 280, 300

With old schema we would get:
```
time = 100
time_end = 300
interval = 20
```
For a total of 64+64+32 = 160 bytes

With new schema we would get:
```
time = 100
interval = 20
count = 10
```
For a total of 64+32+32 = 128 bytes, -20% in size
Nice thing is that: int32 * uint32 = int64
Proof: int32 * uint32 = 2^(32-1) * 2^32 = 2^(31+32) = 2^63 = int64

New schema would be:
```
struct Timeframe {
  time: int64; // seconds since Unix-epoch
  // time_end: int64;
  interval: int32; // seconds
  /// how many intervals?
  count: uint32;
}
```

**Reply by patrick-zippenfenig** [reply]
Hey, sorry for the delay. 

Thanks for the Geocoding API flatbuffers schema. I haven't had time to integrate it, but it looks feasible.

I am still undecided whether to organise all flat buffer types into different files, or keep just one fbs file per "API". I will check again, when I integrate the geocoding API.

I like the idea of using `count` instead of `time_end`, however, the current version is already live and this would be a breaking change. The only solution would be to introduce a `v2`, but this would add a lot of work....

**Reply by DadiBit** [reply]
> Hey, sorry for the delay.
>
> Thanks for the Geocoding API flatbuffers schema. I haven't had time to integrate it, but it looks feasible.

No worries, I saw you were quite busy with pushing commits... Keep up the great work! :+1: 

> I am still undecided whether to organise all flat buffer types into different files, or keep just one fbs file per "API". I will check again, when I integrate the geocoding API.

Coming from mostly OOP languages, I would at least split the Unit and Aggregation enums in separate files, since they are imported pretty much everywhere.

> I like the idea of using count instead of time_end, however, the current version is already live and this would be a breaking change. The only solution would be to introduce a v2, but this would add a lot of work....

The `(deprecated)` flag could help in this situation: it shouldn't break anything, but it would still mess stuff up. Perhaps it could be kept as a sidenote (I mean, it's always 32 bytes even if with 100 variables and 1 million values, not super important) to add in a hypothetical v2.

A bigger/more interesting idea I had, is using an array of `TimeWithVariables` (here I renamed it `Timeframe` while experimenting a bit) instead of single entries for daily/hourly/etc.:
```
// Timeframe containing some variables and relative data over itself
table Timeframe {
  /// The start time of the timeframe
  time: int64;
  /// The end time of the timeframe
  time_end: int64 (deprecated);
  /// Interval (daily, hourly, 15_minutely...) in seconds
  interval: int32;
  /// List of variables with relative values
  variables: [VariableWithValues];
  /// Count of retrieved intervals == values (nulls included) in every variable
  count: uint32;
}

table Response {
  // ...
  current: Timeframe (deprecated);
  daily: Timeframe (deprecated);
  hourly: Timeframe (deprecated);
  minutely_15: Timeframe (deprecated);
  six_hourly: Timeframe (deprecated);
  timeframes: [Timeframe];
}
```
I don't know how much data can be saved, if any, but I feel it's a lot cleaner (and future-proof) to have an array of `Timeframe`s, just like the array of `VariableWithValues`

**Reply by patrick-zippenfenig** [reply]
I also considered an array of `VariableWithValues`. I used individual sections for `hourly`, `daily` or `minutely_15` because it keep the client code simpler. E.g. `result.Hourly().Variables(0)`. 

If it is an array, you need to know the array index: 

```swift
ler hourlyIndex = results.first(where: {$0.time().interval == 3600})
let res = result.Timeframes(hourlyIndex).Variables(0)
```
